From: Jordan Santell <jsantell@mozilla.com>
Date: Fri, 22 May 2015 11:56:49 -0700
Subject: Bug 1049865 - part 2 - fix up browser_profiler_tree-frame-node.js to handle the new stripping of query params from a displayed filename in tree models.

diff --git a/browser/devtools/performance/test/browser_profiler_tree-frame-node.js b/browser/devtools/performance/test/browser_profiler_tree-frame-node.js
index 9f36431..7b35f95 100644
--- a/browser/devtools/performance/test/browser_profiler_tree-frame-node.js
+++ b/browser/devtools/performance/test/browser_profiler_tree-frame-node.js
@@ -43,17 +43,17 @@ function test() {
       location: "hello/<.world (http://foo/bar.js#baz:123:987)"
     })
   }, false);
 
   is(frame2.getInfo().nodeType, "Frame",
     "The second frame node has the correct type.");
   is(frame2.getInfo().functionName, "hello/<.world",
     "The second frame node has the correct function name.");
-  is(frame2.getInfo().fileName, "bar.js#baz",
+  is(frame2.getInfo().fileName, "bar.js",
     "The second frame node has the correct file name.");
   is(frame2.getInfo().hostName, "foo",
     "The second frame node has the correct host name.");
   is(frame2.getInfo().url, "http://foo/bar.js#baz",
     "The second frame node has the correct url.");
   is(frame2.getInfo().line, 123,
     "The second frame node has the correct line.");
   is(frame2.getInfo().column, 987,
@@ -70,17 +70,17 @@ function test() {
       location: "hello/<.world (http://foo/#bar:123:987)"
     })
   }, false);
 
   is(frame3.getInfo().nodeType, "Frame",
     "The third frame node has the correct type.");
   is(frame3.getInfo().functionName, "hello/<.world",
     "The third frame node has the correct function name.");
-  is(frame3.getInfo().fileName, "#bar",
+  is(frame3.getInfo().fileName, "/",
     "The third frame node has the correct file name.");
   is(frame3.getInfo().hostName, "foo",
     "The third frame node has the correct host name.");
   is(frame3.getInfo().url, "http://foo/#bar",
     "The third frame node has the correct url.");
   is(frame3.getInfo().line, 123,
     "The third frame node has the correct line.");
   is(frame3.getInfo().column, 987,
-- 
2.2.1

