From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 4 Aug 2015 18:11:24 -0700
Subject: Bug 1050085 - Clean up performance L10N comments. r=flod


diff --git a/browser/locales/en-US/chrome/browser/devtools/performance.dtd b/browser/locales/en-US/chrome/browser/devtools/performance.dtd
index 5d08db9..56b2202 100644
--- a/browser/locales/en-US/chrome/browser/devtools/performance.dtd
+++ b/browser/locales/en-US/chrome/browser/devtools/performance.dtd
@@ -46,17 +46,17 @@
   -  on a button that opens a dialog to import a saved profile data file. -->
 <!ENTITY performanceUI.importButton "Import…">
 
 <!-- LOCALIZATION NOTE (performanceUI.exportButton): This string is displayed
   -  on a button that opens a dialog to export a saved profile data file. -->
 <!ENTITY performanceUI.exportButton "Save">
 
 <!-- LOCALIZATION NOTE (performanceUI.clearButton): This string is displayed
-  -  on a button that remvoes all the recordings. -->
+  -  on a button that removes all the recordings. -->
 <!ENTITY performanceUI.clearButton "Clear">
 
 <!-- LOCALIZATION NOTE (performanceUI.toolbar.*): These strings are displayed
   -  in the toolbar on buttons that select which view is currently shown. -->
 <!ENTITY performanceUI.toolbar.waterfall "Waterfall">
 <!ENTITY performanceUI.toolbar.js-calltree "Call Tree">
 <!ENTITY performanceUI.toolbar.memory-calltree "Allocations">
 <!ENTITY performanceUI.toolbar.js-flamegraph "JS Flame Chart">
diff --git a/browser/locales/en-US/chrome/browser/devtools/performance.properties b/browser/locales/en-US/chrome/browser/devtools/performance.properties
index 8cd7a70..66857ff 100644
--- a/browser/locales/en-US/chrome/browser/devtools/performance.properties
+++ b/browser/locales/en-US/chrome/browser/devtools/performance.properties
@@ -31,17 +31,18 @@ performance.accesskey=P
 performance.tooltip=Performance (%S)
 
 # LOCALIZATION NOTE (noRecordingsText): The text to display in the
 # recordings menu when there are no recorded profiles yet.
 noRecordingsText=There are no profiles yet.
 
 # LOCALIZATION NOTE (recordingsList.itemLabel):
 # This string is displayed in the recordings list of the Performance Tools,
-# identifying a set of function calls.
+# identifying a set of function calls. %S represents the number of recording,
+# iterating for every new recording, resulting in "Recording #1", "Recording #2", etc.
 recordingsList.itemLabel=Recording #%S
 
 # LOCALIZATION NOTE (recordingsList.recordingLabel):
 # This string is displayed in the recordings list of the Performance Tools,
 # for an item that has not finished recording.
 recordingsList.recordingLabel=In progress…
 
 # LOCALIZATION NOTE (recordingsList.loadingLabel):
-- 
2.2.1

