From b2a787fb763834b329d0ecd2417a6d4f086ce067 Mon Sep 17 00:00:00 2001
From: Jordan Santell <jsantell@gmail.com>
Date: Tue, 16 Sep 2014 16:08:32 -0700
Subject: Bug 1068372 - Disable web audio editor option in browser toolbox. r=vp

---
 browser/devtools/main.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/browser/devtools/main.js b/browser/devtools/main.js
index 795421d..f125094 100644
--- a/browser/devtools/main.js
+++ b/browser/devtools/main.js
@@ -253,17 +253,17 @@ Tools.webAudioEditor = {
   visibilityswitch: "devtools.webaudioeditor.enabled",
   icon: "chrome://browser/skin/devtools/tool-webaudio.svg",
   invertIconForLightTheme: true,
   url: "chrome://browser/content/devtools/webaudioeditor.xul",
   label: l10n("ToolboxWebAudioEditor1.label", webAudioEditorStrings),
   panelLabel: l10n("ToolboxWebAudioEditor1.panelLabel", webAudioEditorStrings),
   tooltip: l10n("ToolboxWebAudioEditor1.tooltip", webAudioEditorStrings),
   isTargetSupported: function(target) {
-    return !target.isAddon;
+    return !target.isAddon && !target.chrome;
   },
   build: function(iframeWindow, toolbox) {
     let panel = new WebAudioEditorPanel(iframeWindow, toolbox);
     return panel.open();
   }
 };
 
 Tools.jsprofiler = {
-- 
1.8.4.2

