From b465e85b7ef96fd1985f9571b8833ef09f64b463 Mon Sep 17 00:00:00 2001
From: Jordan Santell <jsantell@gmail.com>
Date: Tue, 20 Jan 2015 16:01:50 -0800
Subject: Bug 1118497 - Update web audio editor 'parameters' view to
 the more accurately described 'properties' view. r=vp

diff --git a/browser/devtools/webaudioeditor/webaudioeditor.xul b/browser/devtools/webaudioeditor/webaudioeditor.xul
index 68052ec..e60768e 100644
--- a/browser/devtools/webaudioeditor/webaudioeditor.xul
+++ b/browser/devtools/webaudioeditor/webaudioeditor.xul
@@ -90,17 +90,17 @@
                 <hbox class="devtools-toolbarbutton-group">
                   <toolbarbutton class="bypass devtools-toolbarbutton"
                                  data-command="bypass"
                                  tabindex="0"/>
                 </hbox>
               </toolbar>
               <tabs>
                 <tab id="properties-tab"
-                     label="&webAudioEditorUI.tab.properties;"/>
+                     label="&webAudioEditorUI.tab.properties2;"/>
                 <tab id="automation-tab"
                      label="&webAudioEditorUI.tab.automation;"/>
               </tabs>
               <tabpanels flex="1">
                 <!-- Properties Panel -->
                 <tabpanel id="properties-tabpanel"
                           class="tabpanel-content">
                   <vbox id="properties-content" flex="1" hidden="true">
diff --git a/browser/locales/en-US/chrome/browser/devtools/webaudioeditor.dtd b/browser/locales/en-US/chrome/browser/devtools/webaudioeditor.dtd
index 5456443..778c24c 100644
--- a/browser/locales/en-US/chrome/browser/devtools/webaudioeditor.dtd
+++ b/browser/locales/en-US/chrome/browser/devtools/webaudioeditor.dtd
@@ -18,19 +18,19 @@
 <!-- LOCALIZATION NOTE (webAudioEditorUI.reloadNotice2): This is the label shown
   -  along with the button that triggers a page refresh. -->
 <!ENTITY webAudioEditorUI.reloadNotice2   "the page to view and edit the audio context.">
 
 <!-- LOCALIZATION NOTE (webAudioEditorUI.emptyNotice): This is the label shown
   -  while the page is refreshing and the tool waits for a audio context. -->
 <!ENTITY webAudioEditorUI.emptyNotice     "Waiting for an audio context to be created…">
 
-<!-- LOCALIZATION NOTE (webAudioEditorUI.tab.properties): This is the label shown
+<!-- LOCALIZATION NOTE (webAudioEditorUI.tab.properties2): This is the label shown
   -  for the properties tab view. -->
-<!ENTITY webAudioEditorUI.tab.properties  "Parameters">
+<!ENTITY webAudioEditorUI.tab.properties2 "Properties">
 
 <!-- LOCALIZATION NOTE (webAudioEditorUI.tab.automation): This is the label shown
   -  for the automation tab view. -->
 <!ENTITY webAudioEditorUI.tab.automation  "Automation">
 
 <!-- LOCALIZATION NOTE (webAudioEditorUI.inspectorTitle): This is the title for the
   -  AudioNode inspector view. -->
 <!ENTITY webAudioEditorUI.inspectorTitle  "AudioNode Inspector">
diff --git a/browser/themes/shared/devtools/webaudioeditor.inc.css b/browser/themes/shared/devtools/webaudioeditor.inc.css
index 500af7c..bec3c23 100644
--- a/browser/themes/shared/devtools/webaudioeditor.inc.css
+++ b/browser/themes/shared/devtools/webaudioeditor.inc.css
@@ -138,16 +138,22 @@ text {
 .nodes text {
   cursor: pointer;
 }
 
 /**
  * Inspector Styles
  */
 
+/* hide the variables view scope title as its redundant,
+ * because there's only one scope displayed. */
+.variables-view-scope > .title {
+  display: none;
+}
+
 #web-audio-inspector-title {
   margin: 6px;
 }
 
 .web-audio-inspector .error {
   background-image: url(alerticon-warning.png);
   background-size: 13px 12px;
   -moz-appearance: none;
-- 
2.2.1

