From: Jordan Santell <jsantell@mozilla.com>
Date: Sat, 12 Sep 2015 09:09:48 -0700
Subject: Bug 1122058 - Add tests for adding telemetry to performance tool. r=vp,mratcliffe

diff --git a/browser/devtools/performance/test/browser.ini b/browser/devtools/performance/test/browser.ini
index 5455c52d..ec4aaed 100644
--- a/browser/devtools/performance/test/browser.ini
+++ b/browser/devtools/performance/test/browser.ini
@@ -114,16 +114,17 @@ skip-if = os == 'linux' # bug 1186322
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-selected-02.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-selected-03.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-selected-04.js]
 skip-if = os == 'linux' || debug # bug 1186322 for Linux, bug 1203895 for leaks
 [browser_perf-theme-toggle-01.js]
+[browser_perf-telemetry.js]
 [browser_profiler_tree-abstract-01.js]
 [browser_profiler_tree-abstract-02.js]
 [browser_profiler_tree-abstract-03.js]
 [browser_profiler_tree-abstract-04.js]
 [browser_profiler_tree-view-01.js]
 [browser_profiler_tree-view-02.js]
 [browser_profiler_tree-view-03.js]
 [browser_profiler_tree-view-04.js]
-- 
2.2.1

