From: Jordan Santell <jsantell@gmail.com>
Date: Mon, 23 Feb 2015 10:45:40 -0800
Subject: Bug 1132474 - Further clarify restrictions of TabTarget's actor description methods in the comments. r=jryans

diff --git a/browser/devtools/framework/target.js b/browser/devtools/framework/target.js
index 6d0d409..50c7531 100644
--- a/browser/devtools/framework/target.js
+++ b/browser/devtools/framework/target.js
@@ -179,17 +179,23 @@ function TabTarget(tab) {
 
 TabTarget.prototype = {
   _webProgressListener: null,
 
   /**
    * Returns a promise for the protocol description from the root actor.
    * Used internally with `target.actorHasMethod`. Takes advantage of
    * caching if definition was fetched previously with the corresponding
-   * actor information. Must be a remote target.
+   * actor information. Actors are lazily loaded, so not only must the tool using
+   * a specific actor be in use, the actors are only registered after invoking
+   * a method (for performance reasons, added in bug 988237), so to use these actor
+   * detection methods, one must already be communicating with a specific actor of
+   * that type.
+   *
+   * Must be a remote target.
    *
    * @return {Promise}
    * {
    *   "category": "actor",
    *   "typeName": "longstractor",
    *   "methods": [{
    *     "name": "substring",
    *     "request": {
@@ -245,17 +251,18 @@ TabTarget.prototype = {
     if (this.form) {
       return !!this.form[actorName + "Actor"];
     }
     return false;
   },
 
   /**
    * Queries the protocol description to see if an actor has
-   * an available method. The actor must already be lazily-loaded,
+   * an available method. The actor must already be lazily-loaded (read
+   * the restrictions in the `getActorDescription` comments),
    * so this is for use inside of tool. Returns a promise that
    * resolves to a boolean. Must be a remote target.
    *
    * @param {String} actorName
    * @param {String} methodName
    * @return {Promise}
    */
   actorHasMethod: function (actorName, methodName) {
-- 
2.2.1

