From: Jordan Santell <jsantell@gmail.com>
Date: Mon, 23 Feb 2015 10:25:45 -0800
Subject: Bug 1133058 - OptionsView button, when clicked, should have an 'open' attribute. r=vp


diff --git a/browser/devtools/shared/options-view.js b/browser/devtools/shared/options-view.js
index 2ca4d91..b6d4f56 100644
--- a/browser/devtools/shared/options-view.js
+++ b/browser/devtools/shared/options-view.js
@@ -16,16 +16,19 @@ const PREF_CHANGE_EVENT = "pref-changed";
  */
 const OptionsView = function (options={}) {
   this.branchName = options.branchName;
   this.menupopup = options.menupopup;
   this.window = this.menupopup.ownerDocument.defaultView;
   let { document } = this.window;
   this.$ = document.querySelector.bind(document);
   this.$$ = document.querySelectorAll.bind(document);
+  // Get the corresponding button that opens the popup by looking
+  // for an element with a `popup` attribute matching the menu's ID
+  this.button = this.$(`[popup=${this.menupopup.getAttribute("id")}]`);
 
   this.prefObserver = new PrefObserver(this.branchName);
 
   EventEmitter.decorate(this);
 };
 exports.OptionsView = OptionsView;
 
 OptionsView.prototype = {
@@ -121,24 +124,26 @@ OptionsView.prototype = {
     this.prefObserver.set(prefName, value);
   },
 
   /**
    * Fired when the `menupopup` is opened, bound via XUL.
    * Fires an event used in tests.
    */
   _onPopupShown: function () {
+    this.button.setAttribute("open", true);
     this.emit(OPTIONS_SHOWN_EVENT);
   },
 
   /**
    * Fired when the `menupopup` is closed, bound via XUL.
    * Fires an event used in tests.
    */
   _onPopupHidden: function () {
+    this.button.removeAttribute("open");
     this.emit(OPTIONS_HIDDEN_EVENT);
   }
 };
 
 /**
  * Constructor for PrefObserver. Small helper for observing changes
  * on a preference branch. Takes a `branchName`, like "devtools.debugger."
  *
diff --git a/browser/devtools/shared/test/browser_options-view-01.js b/browser/devtools/shared/test/browser_options-view-01.js
index 398b517..b64cce3 100644
--- a/browser/devtools/shared/test/browser_options-view-01.js
+++ b/browser/devtools/shared/test/browser_options-view-01.js
@@ -90,12 +90,14 @@ function createOptionsView(win) {
 
 function* click(view, win, menuitem) {
   let opened = view.once("options-shown");
   let closed = view.once("options-hidden");
 
   let button = win.document.querySelector("#options-button");
   EventUtils.synthesizeMouseAtCenter(button, {}, win);
   yield opened;
+  is(button.getAttribute("open"), "true", "button has `open` attribute");
 
   EventUtils.synthesizeMouseAtCenter(menuitem, {}, win);
   yield closed;
+  ok(!button.hasAttribute("open"), "button does not have `open` attribute");
 }
-- 
2.2.1

