From: Jordan Santell <jsantell@gmail.com>
Date: Thu, 2 Apr 2015 16:17:11 -0700
Subject: Bug 1150733 - Correctly internationalize jit samples label. r=vp,flod

diff --git a/browser/devtools/performance/performance-controller.js b/browser/devtools/performance/performance-controller.js
index a3e52bd..5fdd031 100644
--- a/browser/devtools/performance/performance-controller.js
+++ b/browser/devtools/performance/performance-controller.js
@@ -51,16 +51,18 @@ devtools.lazyImporter(this, "CanvasGraphUtils",
 devtools.lazyImporter(this, "LineGraphWidget",
   "resource:///modules/devtools/Graphs.jsm");
 devtools.lazyImporter(this, "FlameGraphUtils",
   "resource:///modules/devtools/FlameGraph.jsm");
 devtools.lazyImporter(this, "FlameGraph",
   "resource:///modules/devtools/FlameGraph.jsm");
 devtools.lazyImporter(this, "SideMenuWidget",
   "resource:///modules/devtools/SideMenuWidget.jsm");
+devtools.lazyImporter(this, "PluralForm",
+  "resource://gre/modules/PluralForm.jsm");
 
 const BRANCH_NAME = "devtools.performance.ui.";
 
 // Events emitted by various objects in the panel.
 const EVENTS = {
   // Fired by the PerformanceController and OptionsView when a pref changes.
   PREF_CHANGED: "Performance:PrefChanged",
 
diff --git a/browser/devtools/performance/views/jit-optimizations.js b/browser/devtools/performance/views/jit-optimizations.js
index 5a18fe3..6bc68cd 100644
--- a/browser/devtools/performance/views/jit-optimizations.js
+++ b/browser/devtools/performance/views/jit-optimizations.js
@@ -1,16 +1,16 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 "use strict";
 
 const URL_LABEL_TOOLTIP = L10N.getStr("table.url.tooltiptext");
 const OPTIMIZATION_FAILURE = L10N.getStr("jit.optimizationFailure");
-const JIT_SAMPLES = L10N.getStr("jit.samples");
+const JIT_SAMPLES = L10N.getStr("jit.samples2");
 const JIT_EMPTY_TEXT = L10N.getStr("jit.empty");
 
 /**
  * View for rendering JIT Optimization data. The terminology and types
  * used here can be referenced:
  * @see browser/devtools/shared/profiler/jit.js
  */
 
@@ -204,17 +204,18 @@ let JITOptimizationsView = {
     if (!site.hasSuccessfulOutcome()) {
       let icon = document.createElement("span");
       icon.setAttribute("tooltiptext", OPTIMIZATION_FAILURE);
       icon.setAttribute("severity", "warning");
       icon.className = "opt-icon";
       node.appendChild(icon);
     }
 
-    desc.textContent = `${lastStrategy} - (${site.samples} ${JIT_SAMPLES})`;
+    let sampleString = PluralForm.get(site.samples, JIT_SAMPLES).replace("#1", site.samples);
+    desc.textContent = `${lastStrategy} – (${sampleString})`;
     line.textContent = site.data.line;
     line.className = "opt-line";
     column.textContent = site.data.column;
     column.className = "opt-line";
     node.appendChild(desc);
     node.appendChild(urlNode);
     node.appendChild(line);
     node.appendChild(column);
diff --git a/browser/locales/en-US/chrome/browser/devtools/profiler.properties b/browser/locales/en-US/chrome/browser/devtools/profiler.properties
index e014c22..7622b99 100644
--- a/browser/locales/en-US/chrome/browser/devtools/profiler.properties
+++ b/browser/locales/en-US/chrome/browser/devtools/profiler.properties
@@ -120,16 +120,19 @@ recordingsList.saveDialogJSONFilter=JSON Files
 # LOCALIZATION NOTE (recordingsList.saveDialogAllFilter):
 # This string is displayed as a filter for saving a recording to disk.
 recordingsList.saveDialogAllFilter=All Files
 
 # LOCALIZATION NOTE (jit.optimizationFailure):
 # This string is displayed in a tooltip when no JIT optimizations were detected.
 jit.optimizationFailure=Optimization failed
 
-# LOCALIZATION NOTE (jit.samples):
-# This string is displayed for the unit representing thenumber of times a
+# LOCALIZATION NOTE (jit.samples2):
+# See: http://developer.mozilla.org/en/docs/Localization_and_Plurals
+# This string is displayed for the unit representing the number of times a
 # frame is sampled.
-jit.samples=samples
+# "#1" represents the number of samples
+# example: 30 samples
+jit.samples2=#1 sample;#1 samples
 
 # LOCALIZATION NOTE (jit.empty):
 # This string is displayed when there are no JIT optimizations to display.
 jit.empty=No JIT optimizations recorded for this frame.
-- 
2.2.1

