From: Jordan Santell <jsantell@mozilla.com>
Date: Fri, 19 Jun 2015 17:47:13 -0700
Subject: Bug 1150295 pt2: fix rebase error on renaming optimization hint. a=kwierso

diff --git a/browser/devtools/performance/views/details-js-call-tree.js b/browser/devtools/performance/views/details-js-call-tree.js
index 53ea801..eac9110 100644
--- a/browser/devtools/performance/views/details-js-call-tree.js
+++ b/browser/devtools/performance/views/details-js-call-tree.js
@@ -101,17 +101,17 @@ let JsCallTreeView = Heritage.extend(DetailsSubview, {
     let root = new CallView({
       frame: frameNode,
       inverted: inverted,
       // The synthesized root node is hidden in inverted call trees.
       hidden: inverted,
       // Call trees should only auto-expand when not inverted. Passing undefined
       // will default to the CALL_TREE_AUTO_EXPAND depth.
       autoExpandDepth: inverted ? 0 : undefined,
-      enableOptimizations: options.enableOptimizations
+      showOptimizationHint: options.showOptimizationHint
     });
 
     // Bind events.
     root.on("link", this._onLink);
 
     // Pipe "focus" events to the view, mostly for tests
     root.on("focus", () => this.emit("focus"));
     // TODO tests for optimization event and rendering
-- 
2.2.1

