From: Jordan Santell <jsantell@mozilla.com>
Date: Thu, 21 May 2015 23:06:19 -0700
Subject: Bug 1150761 - Rename the performance tool's details view names to better describe the data visualizations. r=vp

diff --git a/browser/devtools/performance/performance.xul b/browser/devtools/performance/performance.xul
index 57c145b..7b0f2b7 100644
--- a/browser/devtools/performance/performance.xul
+++ b/browser/devtools/performance/performance.xul
@@ -99,37 +99,37 @@
           <toolbarbutton id="filter-button"
                          class="devtools-toolbarbutton"
                          popup="performance-filter-menupopup"
                          tooltiptext="&profilerUI.options.filter.tooltiptext;"/>
         </hbox>
         <hbox id="performance-toolbar-controls-detail-views" class="devtools-toolbarbutton-group">
           <toolbarbutton id="select-waterfall-view"
                          class="devtools-toolbarbutton devtools-button"
-                         label="&profilerUI.toolbar.waterfall;"
+                         label="Waterfall"
                          hidden="true"
                          data-view="waterfall" />
           <toolbarbutton id="select-js-calltree-view"
                          class="devtools-toolbarbutton devtools-button"
-                         label="&profilerUI.toolbar.js-calltree;"
+                         label="Call Tree"
                          hidden="true"
                          data-view="js-calltree" />
           <toolbarbutton id="select-js-flamegraph-view"
                          class="devtools-toolbarbutton devtools-button"
-                         label="&profilerUI.toolbar.js-flamegraph;"
+                         label="Flame Chart"
                          hidden="true"
                          data-view="js-flamegraph" />
           <toolbarbutton id="select-memory-calltree-view"
                          class="devtools-toolbarbutton devtools-button"
-                         label="&profilerUI.toolbar.memory-calltree1;"
+                         label="Allocations Tree"
                          hidden="true"
                          data-view="memory-calltree" />
           <toolbarbutton id="select-memory-flamegraph-view"
                          class="devtools-toolbarbutton devtools-button"
-                         label="&profilerUI.toolbar.memory-flamegraph1;"
+                         label="Allocations Chart"
                          hidden="true"
                          data-view="memory-flamegraph" />
         </hbox>
         <spacer flex="1"></spacer>
         <hbox id="performance-toolbar-control-options" class="devtools-toolbarbutton-group">
           <toolbarbutton id="performance-options-button"
                          class="devtools-toolbarbutton devtools-option-toolbarbutton"
                          popup="performance-options-menupopup"
-- 
2.2.1

