From: Jordan Santell <jsantell@gmail.com>
Date: Tue, 28 Apr 2015 20:46:50 -0700
Subject: Bug 1154018 - Check to see that nsIURI's host exists when parsing location for framenodes, and cache failures. r=vp

diff --git a/browser/devtools/shared/profiler/frame-utils.js b/browser/devtools/shared/profiler/frame-utils.js
index 92578a0..0a7dcd1 100644
--- a/browser/devtools/shared/profiler/frame-utils.js
+++ b/browser/devtools/shared/profiler/frame-utils.js
@@ -112,24 +112,30 @@ exports.filterPlatformData = function filterPlatformData (frames) {
   return result;
 }
 
 /**
  * Helper for getting an nsIURL instance out of a string.
  */
 function nsIURL(url) {
   let cached = gNSURLStore.get(url);
-  if (cached) {
+  // If we cached a valid URI, or `null` in the case
+  // of a failure, return it.
+  if (cached !== void 0) {
     return cached;
   }
   let uri = null;
   try {
     uri = Services.io.newURI(url, null, null).QueryInterface(Ci.nsIURL);
+    // Access the host, because the constructor doesn't necessarily throw
+    // if it's invalid, but accessing the host can throw as well
+    uri.host;
   } catch(e) {
     // The passed url string is invalid.
+    uri = null;
   }
   gNSURLStore.set(url, uri);
   return uri;
 }
 
 /**
  * Takes a `host` string from an nsIURL instance and
  * returns the same string, or null, if it's an invalid host.
-- 
2.2.1

