From: Jordan Santell <jsantell@gmail.com>
Date: Thu, 30 Apr 2015 16:03:43 -0700
Subject: Bug 1157523 - Fix intermittent where markers are selected in the waterfall views when there is no recording selected. r=vp

diff --git a/browser/devtools/performance/views/details-waterfall.js b/browser/devtools/performance/views/details-waterfall.js
index 2126131..3f913a3 100644
--- a/browser/devtools/performance/views/details-waterfall.js
+++ b/browser/devtools/performance/views/details-waterfall.js
@@ -69,16 +69,21 @@ let WaterfallView = Heritage.extend(DetailsSubview, {
   },
 
   /**
    * Called when a marker is selected in the waterfall view,
    * updating the markers detail view.
    */
   _onMarkerSelected: function (event, marker) {
     let recording = PerformanceController.getCurrentRecording();
+    // Race condition in tests due to lazy rendering of markers in the
+    // waterfall? intermittent bug 1157523
+    if (!recording) {
+      return;
+    }
     let frames = recording.getFrames();
 
     if (event === "selected") {
       this.details.render({ toolbox: gToolbox, marker, frames });
     }
     if (event === "unselected") {
       this.details.empty();
     }
-- 
2.2.1

