From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 19 May 2015 13:15:37 -0700
Subject: Bug 1157916 - Add tooltip for displaying start/end times on marker details' duration field. r=vp

diff --git a/browser/devtools/performance/test/browser_timeline-waterfall-sidebar.js b/browser/devtools/performance/test/browser_timeline-waterfall-sidebar.js
index d26520b..7c56a5f 100644
--- a/browser/devtools/performance/test/browser_timeline-waterfall-sidebar.js
+++ b/browser/devtools/performance/test/browser_timeline-waterfall-sidebar.js
@@ -37,18 +37,23 @@ function spawnTest () {
   for (let i = 0; i < bars.length; i++) {
     let bar = bars[i];
     bar.click();
     let m = markers[i];
 
     is($("#waterfall-details .marker-details-type").getAttribute("value"), getMarkerLabel(m),
       "sidebar title matches markers name");
 
+    let tooltip = $(".marker-details-duration").getAttribute("tooltiptext");
     let printedDuration = $(".marker-details-duration .marker-details-labelvalue").getAttribute("value");
 
     let toMs = ms => L10N.getFormatStrWithNumbers("timeline.tick", ms);
 
     // Values are rounded. We don't use a strict equality.
     is(toMs(m.end - m.start), printedDuration, "sidebar duration is valid");
+    // For some reason, anything that creates "→" here turns it into a "â" for some reason.
+    // So just check that start and end time are in there somewhere.
+    ok(tooltip.indexOf(toMs(m.start)) !== -1, "tooltip has start time");
+    ok(tooltip.indexOf(toMs(m.end)) !== -1, "tooltip has end time");
   }
   yield teardown(panel);
   finish();
 }
diff --git a/browser/devtools/shared/timeline/marker-utils.js b/browser/devtools/shared/timeline/marker-utils.js
index c1f4b48..a68ee30 100644
--- a/browser/devtools/shared/timeline/marker-utils.js
+++ b/browser/devtools/shared/timeline/marker-utils.js
@@ -104,19 +104,22 @@ const DOM = exports.DOM = {
    * Builds the duration element, like "Duration: 200ms".
    *
    * @param {Document} doc
    * @param {ProfileTimelineMarker} marker
    * @return {Element}
    */
   buildDuration: function (doc, marker) {
     let label = L10N.getStr("timeline.markerDetail.duration");
-    let value = L10N.getFormatStrWithNumbers("timeline.tick", marker.end - marker.start);
-    let el = DOM.buildNameValueLabel(doc, label, value);
+    let start = L10N.getFormatStrWithNumbers("timeline.tick", marker.start);
+    let end = L10N.getFormatStrWithNumbers("timeline.tick", marker.end);
+    let duration = L10N.getFormatStrWithNumbers("timeline.tick", marker.end - marker.start);
+    let el = DOM.buildNameValueLabel(doc, label, duration);
     el.classList.add("marker-details-duration");
+    el.setAttribute("tooltiptext", `${start} → ${end}`);
     return el;
   },
 
   /**
    * Builds labels for name:value pairs. Like "Start: 100ms",
    * "Duration: 200ms", ...
    *
    * @param {Document} doc
-- 
2.2.1

