From: Jordan Santell <jsantell@gmail.com>
Date: Thu, 23 Apr 2015 14:44:59 -0700
Subject: Bug 1157718 - Do not use Array.prototype.includes in production code that leaves nightly in performance tool. r=fitzgen

diff --git a/browser/devtools/performance/modules/front.js b/browser/devtools/performance/modules/front.js
index 6cf302f..a99a030 100644
--- a/browser/devtools/performance/modules/front.js
+++ b/browser/devtools/performance/modules/front.js
@@ -425,17 +425,17 @@ PerformanceActorsConnection.prototype = {
    * @param RecordingModel model
    *        The corresponding RecordingModel that belongs to the recording session wished to stop.
    * @return RecordingModel
    *         Returns the same model, populated with the profiling data.
    */
   stopRecording: Task.async(function*(model) {
     // If model isn't in the PerformanceActorsConnections internal store,
     // then do nothing.
-    if (!this._recordings.includes(model)) {
+    if (this._recordings.indexOf(model) === -1) {
       return;
     }
 
     // Currently there are two ways profiles stop recording. Either manually in the
     // performance tool, or via console.profileEnd. Once a recording is done,
     // we want to deliver the model to the performance tool (either as a return
     // from the PerformanceFront or via `console-profile-end` event) and then
     // remove it from the internal store.
-- 
2.2.1

