From: Jordan Santell <jsantell@mozilla.com>
Date: Thu, 7 May 2015 16:54:49 -0700
Subject: Bug 1160900 - l10n strings for displaying the profiler's buffer status. r=vp

diff --git a/browser/locales/en-US/chrome/browser/devtools/profiler.dtd b/browser/locales/en-US/chrome/browser/devtools/profiler.dtd
index 4580f90..987f0c2 100644
--- a/browser/locales/en-US/chrome/browser/devtools/profiler.dtd
+++ b/browser/locales/en-US/chrome/browser/devtools/profiler.dtd
@@ -16,16 +16,24 @@
 <!ENTITY profilerUI.emptyNotice1    "Click on the">
 <!ENTITY profilerUI.emptyNotice2    "button to start recording JavaScript function calls.">
 
 <!-- LOCALIZATION NOTE (profilerUI.stopNotice1/2): This is the label shown
   -  in the call list view while recording a profile. -->
 <!ENTITY profilerUI.stopNotice1    "Click on the">
 <!ENTITY profilerUI.stopNotice2    "button again to stop profiling.">
 
+<!-- LOCALIZATION NOTE (profilerUI.bufferStatusTooltip): This string
+  -  is displayed as the tooltip for the buffer capacity during a recording. -->
+<!ENTITY profilerUI.bufferStatusTooltip "The profiler stores samples in a circular buffer, and once the buffer reaches the limit for a recording, newer samples begin to overwrite samples at the beginning of the recording.">
+
+<!-- LOCALIZATION NOTE (profilerUI.bufferStatusFull): This string
+  -  is displayed when the profiler's circular buffer has started to overlap. -->
+<!ENTITY profilerUI.bufferStatusFull "The buffer is full. Older samples are now being overwritten.">
+
 <!-- LOCALIZATION NOTE (profilerUI.loadingNotice): This is the label shown
   -  in the call list view while loading a profile. -->
 <!ENTITY profilerUI.loadingNotice "Loading…">
 
 <!-- LOCALIZATION NOTE (profilerUI.recordButton): This string is displayed
   -  on a button that starts a new profile. -->
 <!ENTITY profilerUI.recordButton.tooltip "Record JavaScript function calls.">
 
diff --git a/browser/locales/en-US/chrome/browser/devtools/profiler.properties b/browser/locales/en-US/chrome/browser/devtools/profiler.properties
index b6fb786..a5b61eec 100644
--- a/browser/locales/en-US/chrome/browser/devtools/profiler.properties
+++ b/browser/locales/en-US/chrome/browser/devtools/profiler.properties
@@ -136,8 +136,12 @@ jit.samples2=#1 sample;#1 samples
 # LOCALIZATION NOTE (jit.empty):
 # This string is displayed when there are no JIT optimizations to display.
 jit.empty=No JIT optimizations recorded for this frame.
 
 # LOCALIZATION NOTE (consoleProfile.recordingNotice/stopCommand):
 # These strings are displayed when a recording is in progress, that was started from the console.
 consoleProfile.recordingNotice=Currently recording profile "%S".
 consoleProfile.stopCommand=Stop profiling by typing \"console.profileEnd(\'%S\')\" into the console.
+
+# LOCALIZATION NOTE (profiler.bufferStatus):
+# This string is displayed illustrating how full the profiler's circular buffer is.
+profiler.bufferStatus=Buffer capacity: %S%
-- 
2.2.1

