From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 9 Jun 2015 13:16:12 -0700
Subject: Bug 1161199 part 2: chunk failure when landing first part of
 bug, landing second part. r=vp

diff --git a/browser/devtools/performance/modules/logic/front.js b/browser/devtools/performance/modules/logic/front.js
index a64e39e..b7e0377 100644
--- a/browser/devtools/performance/modules/logic/front.js
+++ b/browser/devtools/performance/modules/logic/front.js
@@ -460,10 +460,10 @@ function getRecordingModelPrefs () {
     withMemory: Services.prefs.getBoolPref("devtools.performance.ui.enable-memory"),
     withTicks: Services.prefs.getBoolPref("devtools.performance.ui.enable-framerate"),
     withAllocations: Services.prefs.getBoolPref("devtools.performance.ui.enable-allocations"),
     allocationsSampleProbability: +Services.prefs.getCharPref("devtools.performance.memory.sample-probability"),
     allocationsMaxLogLength: Services.prefs.getIntPref("devtools.performance.memory.max-log-length")
   };
 }
 
-exports.getPerformanceActorsConnection = t => SharedPerformanceActors.forTarget(t);
+exports.getPerformanceFront = t => PerformanceFronts.forTarget(t);
 exports.PerformanceFront = PerformanceFront;
-- 
2.2.1

