From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 5 May 2015 10:27:08 -0700
Subject: Bug 1161623 - Use MOZ_DEV_EDITION for displaying non-profiler components in performance tool for fx40.0, instead of channel. r=bgrins

diff --git a/browser/app/profile/firefox.js b/browser/app/profile/firefox.js
index 00b8d3c..55494d6 100644
--- a/browser/app/profile/firefox.js
+++ b/browser/app/profile/firefox.js
@@ -1438,26 +1438,25 @@ pref("devtools.performance.profiler.sample-frequency-khz", 1);
 pref("devtools.performance.ui.invert-call-tree", true);
 pref("devtools.performance.ui.invert-flame-graph", false);
 pref("devtools.performance.ui.flatten-tree-recursion", true);
 pref("devtools.performance.ui.show-platform-data", false);
 pref("devtools.performance.ui.show-idle-blocks", true);
 pref("devtools.performance.ui.enable-memory", false);
 pref("devtools.performance.ui.enable-framerate", true);
 pref("devtools.performance.ui.show-jit-optimizations", false);
-// If in aurora (40.0, will revert for 40.1), set default
+// If in aurora/dev edition (40.0, will revert for 40.1), set default
 // to retro mode.
 // TODO bug 1160313
-#if MOZ_UPDATE_CHANNEL == aurora
+#ifdef MOZ_DEV_EDITION
   pref("devtools.performance.ui.retro-mode", true);
 #else
   pref("devtools.performance.ui.retro-mode", false);
 #endif
 
-
 // The default cache UI setting
 pref("devtools.cache.disabled", false);
 
 // The default service workers UI setting
 pref("devtools.serviceWorkers.testing.enabled", false);
 
 // Enable the Network Monitor
 pref("devtools.netmonitor.enabled", true);
-- 
2.2.1

