From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 4 Aug 2015 17:12:34 -0700
Subject: Bug 1163763 - part 2 - fix broken test from bug 1163763. r=fitzgen. CLOSED TREE

diff --git a/browser/devtools/performance/test/unit/test_marker-utils.js b/browser/devtools/performance/test/unit/test_marker-utils.js
index f9e3d1e..61004df 100644
--- a/browser/devtools/performance/test/unit/test_marker-utils.js
+++ b/browser/devtools/performance/test/unit/test_marker-utils.js
@@ -42,30 +42,30 @@ add_task(function () {
   equal(Utils.getMarkerFields({ name: "Javascript", causeName: "Some Platform Field" })[0].value, "(Gecko)",
     "Correctly obfuscates JS markers when platform data is off.");
   Services.prefs.setBoolPref(PLATFORM_DATA_PREF, true);
   equal(Utils.getMarkerFields({ name: "Javascript", causeName: "Some Platform Field" })[0].value, "Some Platform Field",
     "Correctly deobfuscates JS markers when platform data is on.");
 
   equal(Utils.getMarkerClassName("Javascript"), "Function Call",
     "getMarkerClassName() returns correct string when defined via function");
-  equal(Utils.getMarkerClassName("GarbageCollection"), "GC Event",
+  equal(Utils.getMarkerClassName("GarbageCollection"), "Incremental GC",
     "getMarkerClassName() returns correct string when defined via function");
   equal(Utils.getMarkerClassName("Reflow"), "Layout",
     "getMarkerClassName() returns correct string when defined via string");
 
   TIMELINE_BLUEPRINT["fakemarker"] = { group: 0 };
   try {
     Utils.getMarkerClassName("fakemarker");
     ok(false, "getMarkerClassName() should throw when no label on blueprint.");
   } catch (e) {
     ok(true, "getMarkerClassName() should throw when no label on blueprint.");
   }
 
-  TIMELINE_BLUEPRINT["fakemarker"] = { group: 0, label: () => void 0};
+  TIMELINE_BLUEPRINT["fakemarker"] = { group: 0, label: () => void 0 };
   try {
     Utils.getMarkerClassName("fakemarker");
     ok(false, "getMarkerClassName() should throw when label function returnd undefined.");
   } catch (e) {
     ok(true, "getMarkerClassName() should throw when label function returnd undefined.");
   }
 
   delete TIMELINE_BLUEPRINT["fakemarker"];
-- 
2.2.1

