From: Jordan Santell <jsantell@mozilla.com>
Date: Sat, 9 May 2015 16:05:59 -0700
Subject: Bug 1163268 - Remove UI options for recording with memory and jit optimizations. r=vp

diff --git a/browser/devtools/performance/performance.xul b/browser/devtools/performance/performance.xul
index acd23a4..30c8fc7 100644
--- a/browser/devtools/performance/performance.xul
+++ b/browser/devtools/performance/performance.xul
@@ -32,21 +32,22 @@
   <popupset id="performance-options-popupset">
     <menupopup id="performance-filter-menupopup"/>
     <menupopup id="performance-options-menupopup">
       <menuitem id="option-show-platform-data"
                 type="checkbox"
                 data-pref="show-platform-data"
                 label="&profilerUI.showPlatformData;"
                 tooltiptext="&profilerUI.showPlatformData.tooltiptext;"/>
-      <menuitem id="option-enable-memory"
+      <!-- TODO Re-enable in bug 1163350 -->
+      <!-- <menuitem id="option-enable-memory"
                 type="checkbox"
                 data-pref="enable-memory"
                 label="&profilerUI.enableMemory;"
-                tooltiptext="&profilerUI.enableMemory.tooltiptext;"/>
+                tooltiptext="&profilerUI.enableMemory.tooltiptext;"/> -->
       <menuitem id="option-enable-framerate"
                 type="checkbox"
                 data-pref="enable-framerate"
                 label="&profilerUI.enableFramerate;"
                 tooltiptext="&profilerUI.enableFramerate.tooltiptext;"/>
       <menuitem id="option-invert-call-tree"
                 type="checkbox"
                 data-pref="invert-call-tree"
@@ -57,21 +58,22 @@
                 data-pref="invert-flame-graph"
                 label="&profilerUI.invertFlameGraph;"
                 tooltiptext="&profilerUI.invertFlameGraph.tooltiptext;"/>
       <menuitem id="option-flatten-tree-recursion"
                 type="checkbox"
                 data-pref="flatten-tree-recursion"
                 label="&profilerUI.flattenTreeRecursion;"
                 tooltiptext="&profilerUI.flattenTreeRecursion.tooltiptext;"/>
-      <menuitem id="option-show-jit-optimizations"
+      <!-- TODO Re-enable in bug 1163351 -->
+      <!-- <menuitem id="option-show-jit-optimizations"
                 type="checkbox"
                 data-pref="show-jit-optimizations"
                 label="&profilerUI.showJITOptimizations;"
-                tooltiptext="&profilerUI.showJITOptimizations.tooltiptext;"/>
+                tooltiptext="&profilerUI.showJITOptimizations.tooltiptext;"/> -->
     </menupopup>
   </popupset>
 
   <hbox class="theme-body" flex="1">
     <vbox id="recordings-pane">
       <toolbar id="recordings-toolbar"
                class="devtools-toolbar">
         <hbox id="recordings-controls"
diff --git a/browser/devtools/performance/views/toolbar.js b/browser/devtools/performance/views/toolbar.js
index 05770b8..5c5c024 100644
--- a/browser/devtools/performance/views/toolbar.js
+++ b/browser/devtools/performance/views/toolbar.js
@@ -21,18 +21,23 @@ let ToolbarView = {
       menupopup: $("#performance-options-menupopup")
     });
 
     // TODO bug 1160313 get rid of retro mode checks
     // hide option buttons here, and any other buttons in the toolbar
     // (details.js takes care of view buttons)
     if (PerformanceController.getOption("retro-mode")) {
       let RETRO_ELEMENTS = [
-        "#option-flatten-tree-recursion", "#option-enable-memory", "#option-invert-flame-graph",
-        "#option-show-jit-optimizations", "#filter-button"
+        "#option-flatten-tree-recursion",
+        "#option-invert-flame-graph",
+        "#filter-button",
+        // TODO re-enable in bug 1163350
+        // "#option-enable-memory",
+        // TODO re-enable in bug 1163351
+        // "#option-show-jit-optimizations",
       ];
       for (let selector of RETRO_ELEMENTS) {
         $(selector).hidden = true;
       }
     }
 
     yield this.optionsView.initialize();
     this.optionsView.on("pref-changed", this._onPrefChanged);
-- 
2.2.1

