From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 19 May 2015 15:42:36 -0700
Subject: Bug 1166124 - Change 'buffer capacity' string to a more accurate 'buffer x% full' stirng. r=fitzgen

diff --git a/browser/devtools/performance/performance-view.js b/browser/devtools/performance/performance-view.js
index a9897d5..2d0c0b1 100644
--- a/browser/devtools/performance/performance-view.js
+++ b/browser/devtools/performance/performance-view.js
@@ -166,17 +166,17 @@ let PerformanceView = {
     // Be a little flexible on the buffer status, although not sure how
     // this could happen, as RecordingModel clamps.
     if (percent >= 99) {
       $container.setAttribute("buffer-status", "full");
     } else {
       $container.setAttribute("buffer-status", "in-progress");
     }
 
-    $bufferLabel.value = L10N.getFormatStr("profiler.bufferStatus", percent);
+    $bufferLabel.value = `Buffer ${percent}% full`;
     this.emit(EVENTS.UI_BUFFER_UPDATED, percent);
   },
 
   /**
    * Adds the `locked` attribute on the record button. This prevents it
    * from being clicked while recording is started or stopped.
    */
   _lockRecordButtons: function () {
-- 
2.2.1

