From: Jordan Santell <jsantell@mozilla.com>
Date: Thu, 21 May 2015 22:43:52 -0700
Subject: Bug 1166494 - part2: correctly add labels to console markers and properly add style markers' restyleHints. r=vp

diff --git a/browser/devtools/shared/timeline/global.js b/browser/devtools/shared/timeline/global.js
index 73e8d3f..d142436 100644
--- a/browser/devtools/shared/timeline/global.js
+++ b/browser/devtools/shared/timeline/global.js
@@ -45,17 +45,18 @@ const L10N = new ViewHelpers.L10N(STRINGS_URI);
  * Whenever this is changed, browser_timeline_waterfall-styles.js *must* be
  * updated as well.
  */
 const TIMELINE_BLUEPRINT = {
   /* Group 0 - Reflow and Rendering pipeline */
   "Styles": {
     group: 0,
     colorName: "graphs-purple",
-    label: L10N.getStr("timeline.label.styles2")
+    label: L10N.getStr("timeline.label.styles2"),
+    fields: getStylesFields,
   },
   "Reflow": {
     group: 0,
     colorName: "graphs-purple",
     label: L10N.getStr("timeline.label.reflow2")
   },
   "Paint": {
     group: 0,
@@ -101,26 +102,30 @@ const TIMELINE_BLUEPRINT = {
       { property: "nonincrementalReason", label: "Non-incremental Reason:" }
     ]
   },
 
   /* Group 2 - User Controlled */
   "ConsoleTime": {
     group: 2,
     colorName: "graphs-grey",
-    label: L10N.getStr("timeline.label.consoleTime"),
+    label: sublabelForProperty(L10N.getStr("timeline.label.consoleTime"), "causeName"),
     fields: [{
       property: "causeName",
       label: L10N.getStr("timeline.markerDetail.consoleTimerName")
     }]
   },
   "TimeStamp": {
     group: 2,
     colorName: "graphs-blue",
-    label: L10N.getStr("timeline.label.timestamp")
+    label: sublabelForProperty(L10N.getStr("timeline.label.timestamp"), "causeName"),
+    fields: [{
+      property: "causeName",
+      label: "Label:"
+    }]
   },
 };
 
 /**
  * A series of formatters used by the blueprint.
  */
 
 function getEventPhaseName (marker) {
-- 
2.2.1

