From: Jordan Santell <jsantell@mozilla.com>
Date: Thu, 28 May 2015 14:13:22 -0700
Subject: Bug 1167835 - Performance tools should be on by default when a platform has e10s capabilities, but hidden otherwise. r=vp

diff --git a/browser/app/profile/firefox.js b/browser/app/profile/firefox.js
index 489916f..4d69ea3 100644
--- a/browser/app/profile/firefox.js
+++ b/browser/app/profile/firefox.js
@@ -1436,18 +1436,23 @@ pref("devtools.debugger.tracer", false);
 // The default Debugger UI settings
 pref("devtools.debugger.ui.panes-sources-width", 200);
 pref("devtools.debugger.ui.panes-instruments-width", 300);
 pref("devtools.debugger.ui.panes-visible-on-startup", false);
 pref("devtools.debugger.ui.variables-sorting-enabled", true);
 pref("devtools.debugger.ui.variables-only-enum-visible", false);
 pref("devtools.debugger.ui.variables-searchbox-visible", false);
 
-// Enable the Performance tools
+// Enable the Performance tools by default, only where
+// e10s is an option
+#ifdef E10S_TESTING_ONLY
 pref("devtools.performance.enabled", true);
+#else
+pref("devtools.performance.enabled", false);
+#endif
 
 // The default Performance UI settings
 pref("devtools.performance.memory.sample-probability", "0.05");
 // Can't go higher than this without causing internal allocation overflows while
 // serializing the allocations data over the RDP.
 pref("devtools.performance.memory.max-log-length", 125000);
 pref("devtools.performance.timeline.hidden-markers", "[]");
 pref("devtools.performance.profiler.buffer-size", 10000000);
-- 
2.2.1

