From: Jordan Santell <jsantell@mozilla.com>
Date: Thu, 16 Jul 2015 15:40:23 -0700
Subject: Bug 1172408 - When checking if a marker details element has an action on click, pass in the marker details container so that we don't ascend up to the document, throwing an error. r=vp

diff --git a/browser/devtools/performance/modules/widgets/marker-details.js b/browser/devtools/performance/modules/widgets/marker-details.js
index 12780ea..277546c 100644
--- a/browser/devtools/performance/modules/widgets/marker-details.js
+++ b/browser/devtools/performance/modules/widgets/marker-details.js
@@ -97,17 +97,17 @@ MarkerDetails.prototype = {
   },
 
   /**
    * Handles click in the marker details view. Based on the target,
    * can handle different actions -- only supporting view source links
    * for the moment.
    */
   _onClick: function (e) {
-    let data = findActionFromEvent(e.target);
+    let data = findActionFromEvent(e.target, this._parent);
     if (!data) {
       return;
     }
 
     if (data.action === "view-source") {
       this.emit("view-source", data.url, data.line);
     }
   },
@@ -116,17 +116,17 @@ MarkerDetails.prototype = {
    * Handles the "mouseup" event on the marker details view splitter.
    */
   _onSplitterMouseUp: function() {
     this.emit("resize");
   }
 };
 
 /**
- * Take an element from an event `target`, and asend through
+ * Take an element from an event `target`, and ascend through
  * the DOM, looking for an element with a `data-action` attribute. Return
  * the parsed `data-action` value found, or null if none found before
  * reaching the parent `container`.
  *
  * @param {Element} target
  * @param {Element} container
  * @return {?object}
  */
-- 
2.2.1

