From 6615e770e8c00ee4f4dfb3fbdd5d12e2032c9161 Mon Sep 17 00:00:00 2001
From: Nounours Heureux <schtroumps31@gmail.com>
Date: Tue, 6 Oct 2015 15:48:00 -0700
Subject: Bug 1173775 - Right-clicking a file in the call tree doesn't open it anymore. r=jsantell

diff --git a/devtools/client/performance/modules/widgets/tree-view.js b/devtools/client/performance/modules/widgets/tree-view.js
index 79a58fb..c5babab 100644
--- a/devtools/client/performance/modules/widgets/tree-view.js
+++ b/devtools/client/performance/modules/widgets/tree-view.js
@@ -371,13 +371,16 @@ CallView.prototype = Heritage.extend(AbstractTreeItem.prototype, {
   },
 
   /**
    * Handler for the "click" event on the url node of this call view.
    */
   _onUrlClick: function(e) {
     e.preventDefault();
     e.stopPropagation();
-    this.root.emit("link", this);
+    // Only emit for left click events
+    if (e.button === 0) {
+      this.root.emit("link", this);
+    }
   },
 });
 
 exports.CallView = CallView;
-- 
2.2.1

