From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 6 Oct 2015 15:50:19 -0700
Subject: Bug 1173775 - Tests for bug 1173775. r=vporof

diff --git a/devtools/client/performance/test/browser_profiler_tree-view-06.js b/devtools/client/performance/test/browser_profiler_tree-view-06.js
index ffeb225..1bbc368 100644
--- a/devtools/client/performance/test/browser_profiler_tree-view-06.js
+++ b/devtools/client/performance/test/browser_profiler_tree-view-06.js
@@ -17,22 +17,38 @@ function* spawnTest() {
 
   let container = document.createElement("vbox");
   treeRoot.attachTo(container);
 
   let A = treeRoot.getChild();
   let B = A.getChild();
   let D = B.getChild();
 
-  let receivedLinkEvent = treeRoot.once("link");
+  let linkEvent = null;
+  let handler = (_, e) => linkEvent = e;
+
+  treeRoot.on("link", handler);
+
+  // Fire the right click
+  EventUtils.synthesizeMouseAtCenter(D.target.querySelector(".call-tree-url"), { button: 2, type: "mousedown" }, window);
+
+  // Wait 200ms
+  yield idleWait(50);
+
+  // Ensure link was not called for right click
+  ok(!linkEvent, "`link` event not fired for right click")
+
+  // Fire left click
   EventUtils.sendMouseEvent({ type: "mousedown" }, D.target.querySelector(".call-tree-url"));
 
-  let eventItem = yield receivedLinkEvent;
-  is(eventItem, D, "The 'link' event target is correct.");
+  // Wait until linkEvent is true, should occur quickly, if not synchronously
+  yield waitUntil(() => linkEvent);
+  is(linkEvent, D, "The 'link' event target is correct.");
 
+  treeRoot.off("link", handler);
   finish();
 }
 
 var gThread = synthesizeProfileForTest([{
   time: 5,
   frames: [
     { category: 8,  location: "(root)" },
     { category: 8,  location: "A (http://foo/bar/baz:12)" },
-- 
2.2.1

