From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 4 Aug 2015 18:40:22 -0700
Subject: Bug 1191158 - Promote 'Record Allocations' options to no longer be experimental in the performance tool. r=fitzgen

diff --git a/browser/devtools/performance/performance.xul b/browser/devtools/performance/performance.xul
index a019242..67f38ae 100644
--- a/browser/devtools/performance/performance.xul
+++ b/browser/devtools/performance/performance.xul
@@ -40,17 +40,16 @@
                 tooltiptext="&performanceUI.showPlatformData.tooltiptext;"/>
       <menuitem id="option-enable-memory"
                 class="experimental-option"
                 type="checkbox"
                 data-pref="enable-memory"
                 label="&performanceUI.enableMemory;"
                 tooltiptext="&performanceUI.enableMemory.tooltiptext;"/>
       <menuitem id="option-enable-allocations"
-                class="experimental-option"
                 type="checkbox"
                 data-pref="enable-allocations"
                 label="&performanceUI.enableAllocations;"
                 tooltiptext="&performanceUI.enableAllocations.tooltiptext;"/>
       <menuitem id="option-enable-framerate"
                 type="checkbox"
                 data-pref="enable-framerate"
                 label="&performanceUI.enableFramerate;"
-- 
2.2.1

