From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 8 Sep 2015 14:24:38 -0700
Subject: Bug 1191158 - Revert commit c0764a6b88f5. r=fitzgen

diff --git a/browser/devtools/performance/performance.xul b/browser/devtools/performance/performance.xul
index 1d89129..8420e90 100644
--- a/browser/devtools/performance/performance.xul
+++ b/browser/devtools/performance/performance.xul
@@ -38,16 +38,17 @@
                 tooltiptext="&performanceUI.showPlatformData.tooltiptext;"/>
       <menuitem id="option-enable-memory"
                 class="experimental-option"
                 type="checkbox"
                 data-pref="enable-memory"
                 label="&performanceUI.enableMemory;"
                 tooltiptext="&performanceUI.enableMemory.tooltiptext;"/>
       <menuitem id="option-enable-allocations"
+                class="experimental-option"
                 type="checkbox"
                 data-pref="enable-allocations"
                 label="&performanceUI.enableAllocations;"
                 tooltiptext="&performanceUI.enableAllocations.tooltiptext;"/>
       <menuitem id="option-enable-framerate"
                 type="checkbox"
                 data-pref="enable-framerate"
                 label="&performanceUI.enableFramerate;"
-- 
2.2.1

