From: Jordan Santell <jsantell@mozilla.com>
Date: Mon, 10 Aug 2015 18:43:19 -0700
Subject: Bug 1193110 - Do not close toolbox without waiting for PerformanceActor to finish connecting and tearing down. r=jryans


diff --git a/browser/devtools/framework/toolbox.js b/browser/devtools/framework/toolbox.js
index e94b8a3..29f3f63 100644
--- a/browser/devtools/framework/toolbox.js
+++ b/browser/devtools/framework/toolbox.js
@@ -406,25 +406,23 @@ Toolbox.prototype = {
       yield promise.all([
         splitConsolePromise,
         buttonsPromise,
         framesPromise
       ]);
 
       // Lazily connect to the profiler here and don't wait for it to complete,
       // used to intercept console.profile calls before the performance tools are open.
-      let profilerReady = this.initPerformance();
+      let performanceFrontConnection = this.initPerformance();
 
-      // However, while testing, we must wait for the performance connection to
-      // finish, as most tests shut down without waiting for a toolbox
-      // destruction event, resulting in the shared profiler connection being
-      // opened and closed outside of the test that originally opened the
-      // toolbox.
+      // If in testing environment, wait for performance connection to finish,
+      // so we don't have to explicitly wait for this in tests; ideally, all tests
+      // will handle this on their own, but each have their own tear down function.
       if (DevToolsUtils.testing) {
-        yield profilerReady;
+        yield performanceFrontConnection;
       }
 
       this.emit("ready");
     }.bind(this)).then(null, console.error.bind(console));
   },
 
   _pingTelemetry: function() {
     this._telemetry.toolOpened("toolbox");
@@ -1981,38 +1979,47 @@ Toolbox.prototype = {
    */
   initPerformance: Task.async(function*() {
     // If target does not have profiler actor (addons), do not
     // even register the shared performance connection.
     if (!this.target.hasActor("profiler")) {
       return;
     }
 
-    if (this.performance) {
-      yield this.performance.open();
-      return this.performance;
+    if (this._performanceFrontConnection) {
+      return this._performanceFrontConnection.promise;
     }
 
-    this._performance = getPerformanceFront(this.target);
+    this._performanceFrontConnection = promise.defer();
+
+    this._performance = getPerformanceFront(this._target);
+
     yield this.performance.open();
+
     // Emit an event when connected, but don't wait on startup for this.
     this.emit("profiler-connected");
 
-    return this.performance;
+    this._performanceFrontConnection.resolve(this.performance);
+    return this._performanceFrontConnection.promise;
   }),
 
   /**
    * Disconnects the underlying Performance actor. If the connection
    * has not finished initializing, as opening a toolbox does not wait,
    * the performance connection destroy method will wait for it on its own.
    */
   destroyPerformance: Task.async(function*() {
     if (!this.performance) {
       return;
     }
+    // If still connecting to performance actor, allow the
+    // actor to resolve its connection before attempting to destroy.
+    if (this._performanceFrontConnection) {
+      yield this._performanceFrontConnection.promise;
+    }
     yield this.performance.destroy();
     this._performance = null;
   }),
 
   /**
    * Returns gViewSourceUtils for viewing source.
    */
   get gViewSourceUtils() {
diff --git a/browser/devtools/performance/test/head.js b/browser/devtools/performance/test/head.js
index fa0be18..d13572a 100644
--- a/browser/devtools/performance/test/head.js
+++ b/browser/devtools/performance/test/head.js
@@ -217,16 +217,20 @@ function initPerformance(aUrl, tool="performance", targetOps={}) {
     // may not exist. Possible options that will actually work:
     // TEST_MOCK_MEMORY_ACTOR = true
     // TEST_MOCK_TIMELINE_ACTOR = true
     // TEST_MOCK_PROFILER_CHECK_TIMER = number
     // TEST_PROFILER_FILTER_STATUS = array
     merge(target, targetOps);
 
     let toolbox = yield gDevTools.showToolbox(target, tool);
+
+    // Wait for the performance tool to be spun up
+    yield toolbox.initPerformance();
+
     let panel = toolbox.getCurrentPanel();
     return { target, panel, toolbox };
   });
 }
 
 /**
  * Initializes a webconsole panel. Returns a target, panel and toolbox reference.
  * Also returns a console property that allows calls to `profile` and `profileEnd`.
-- 
2.2.1

