From: Jordan Santell <jsantell@mozilla.com>
Date: Wed, 9 Sep 2015 15:50:44 -0700
Subject: Bug 1194457 - Rename some perftools tests to match naming scheme of other tests. r=vp

diff --git a/browser/devtools/performance/test/browser.ini b/browser/devtools/performance/test/browser.ini
index b07fe62..98ef610 100644
--- a/browser/devtools/performance/test/browser.ini
+++ b/browser/devtools/performance/test/browser.ini
@@ -90,27 +90,27 @@ skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-notices-02.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-notices-03.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-notices-04.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-notices-05.js]
 skip-if = os == 'linux' # bug 1186322
-[browser_perf_recordings-io-01.js]
+[browser_perf-recordings-io-01.js]
 skip-if = os == 'linux' # bug 1186322
-[browser_perf_recordings-io-02.js]
+[browser_perf-recordings-io-02.js]
 skip-if = os == 'linux' # bug 1186322
-[browser_perf_recordings-io-03.js]
+[browser_perf-recordings-io-03.js]
 skip-if = os == 'linux' # bug 1186322
-[browser_perf_recordings-io-04.js]
+[browser_perf-recordings-io-04.js]
 skip-if = os == 'linux' # bug 1186322
-[browser_perf_recordings-io-05.js]
+[browser_perf-recordings-io-05.js]
 skip-if = os == 'linux' # bug 1186322
-[browser_perf_recordings-io-06.js]
+[browser_perf-recordings-io-06.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-range-changed-render.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-selected-01.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-selected-02.js]
 skip-if = os == 'linux' # bug 1186322
 [browser_perf-recording-selected-03.js]
diff --git a/browser/devtools/performance/test/browser_perf_recordings-io-01.js b/browser/devtools/performance/test/browser_perf-recordings-io-01.js
similarity index 100%
rename from browser/devtools/performance/test/browser_perf_recordings-io-01.js
rename to browser/devtools/performance/test/browser_perf-recordings-io-01.js
diff --git a/browser/devtools/performance/test/browser_perf_recordings-io-02.js b/browser/devtools/performance/test/browser_perf-recordings-io-02.js
similarity index 100%
rename from browser/devtools/performance/test/browser_perf_recordings-io-02.js
rename to browser/devtools/performance/test/browser_perf-recordings-io-02.js
diff --git a/browser/devtools/performance/test/browser_perf_recordings-io-03.js b/browser/devtools/performance/test/browser_perf-recordings-io-03.js
similarity index 100%
rename from browser/devtools/performance/test/browser_perf_recordings-io-03.js
rename to browser/devtools/performance/test/browser_perf-recordings-io-03.js
diff --git a/browser/devtools/performance/test/browser_perf_recordings-io-04.js b/browser/devtools/performance/test/browser_perf-recordings-io-04.js
similarity index 100%
rename from browser/devtools/performance/test/browser_perf_recordings-io-04.js
rename to browser/devtools/performance/test/browser_perf-recordings-io-04.js
diff --git a/browser/devtools/performance/test/browser_perf_recordings-io-05.js b/browser/devtools/performance/test/browser_perf-recordings-io-05.js
similarity index 100%
rename from browser/devtools/performance/test/browser_perf_recordings-io-05.js
rename to browser/devtools/performance/test/browser_perf-recordings-io-05.js
diff --git a/browser/devtools/performance/test/browser_perf_recordings-io-06.js b/browser/devtools/performance/test/browser_perf-recordings-io-06.js
similarity index 100%
rename from browser/devtools/performance/test/browser_perf_recordings-io-06.js
rename to browser/devtools/performance/test/browser_perf-recordings-io-06.js
-- 
2.2.1

