From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 26 Jan 2016 22:57:28 -0800
Subject: Bug 1204174 - Reduce test iterations to reduce intermittent timeouts. r=vp

diff --git a/devtools/client/performance/test/browser_perf-overview-render-02.js b/devtools/client/performance/test/browser_perf-overview-render-02.js
index 26a9426..3670917 100644
--- a/devtools/client/performance/test/browser_perf-overview-render-02.js
+++ b/devtools/client/performance/test/browser_perf-overview-render-02.js
@@ -1,15 +1,16 @@
 /* Any copyright is dedicated to the Public Domain.
    http://creativecommons.org/publicdomain/zero/1.0/ */
 
 /**
  * Tests that the overview graphs cannot be selected during recording
  * and that they're cleared upon rerecording.
  */
+const TIMES_TO_UPDATE = 2;
 function* spawnTest() {
   // This test seems to take a long time to cleanup on Ubuntu VMs.
   requestLongerTimeout(2);
 
   let { panel } = yield initPerformance(SIMPLE_URL);
   let { EVENTS, OverviewView } = panel.panelWin;
 
   // Enable memory to test all the overview graphs.
@@ -46,17 +47,17 @@ function* spawnTest() {
   is(memory.selectionEnabled, false,
     "The selection should not be enabled for the memory overview (2).");
   is(memory.hasSelection(), false,
     "The memory overview shouldn't have a selection before recording.");
 
   let updated = 0;
   OverviewView.on(EVENTS.OVERVIEW_RENDERED, () => updated++);
 
-  ok((yield waitUntil(() => updated > 10)),
+  ok((yield waitUntil(() => updated > TIMES_TO_UPDATE)),
     "The overviews were updated several times.");
 
   ok("selectionEnabled" in framerate,
     "The selection should still not be enabled for the framerate overview (1).");
   is(framerate.selectionEnabled, false,
     "The selection should still not be enabled for the framerate overview (2).");
   is(framerate.hasSelection(), false,
     "The framerate overview still shouldn't have a selection before recording.");
diff --git a/devtools/client/performance/test/browser_perf-overview-render-03.js b/devtools/client/performance/test/browser_perf-overview-render-03.js
index a203af7..8fc50ef 100644
--- a/devtools/client/performance/test/browser_perf-overview-render-03.js
+++ b/devtools/client/performance/test/browser_perf-overview-render-03.js
@@ -1,14 +1,15 @@
 /* Any copyright is dedicated to the Public Domain.
    http://creativecommons.org/publicdomain/zero/1.0/ */
 
 /**
  * Tests that the overview graphs share the exact same width and scaling.
  */
+const TIMES_TO_UPDATE = 2;
 function* spawnTest() {
   // This test seems to take a long time to cleanup on Ubuntu VMs.
   requestLongerTimeout(2);
 
   let { panel } = yield initPerformance(SIMPLE_URL);
   let { EVENTS, PerformanceController, OverviewView } = panel.panelWin;
 
   // Enable memory to test all the overview graphs.
@@ -18,17 +19,17 @@ function* spawnTest() {
 
   let updated = 0;
   OverviewView.on(EVENTS.OVERVIEW_RENDERED, () => updated++);
 
   yield busyWait(100);
   yield waitUntil(() => PerformanceController.getCurrentRecording().getMarkers().length);
   yield waitUntil(() => PerformanceController.getCurrentRecording().getMemory().length);
   yield waitUntil(() => PerformanceController.getCurrentRecording().getTicks().length);
-  yield waitUntil(() => updated > 10);
+  yield waitUntil(() => updated > TIMES_TO_UPDATE);
 
   yield stopRecording(panel);
 
   let markers = OverviewView.graphs.get("timeline");
   let framerate = OverviewView.graphs.get("framerate");
   let memory = OverviewView.graphs.get("memory");
 
   ok(markers.width > 0,
-- 
2.3.1

