From: Jordan Santell <jsantell@mozilla.com>
Date: Wed, 28 Oct 2015 07:47:53 -0700
Subject: Bug 1217969 - Highlight the memory tool tab when recording allocations. r=fitzgen

diff --git a/devtools/client/memory/initializer.js b/devtools/client/memory/initializer.js
index 8b62455..35eb7b2 100644
--- a/devtools/client/memory/initializer.js
+++ b/devtools/client/memory/initializer.js
@@ -14,21 +14,44 @@ const { Provider } = require("devtools/client/shared/vendor/react-redux");
 const App = createFactory(require("devtools/client/memory/app"));
 const Store = require("devtools/client/memory/store");
 
 /**
  * The current target, toolbox, MemoryFront, and HeapAnalysesClient, set by this tool's host.
  */
 var gToolbox, gTarget, gFront, gHeapAnalysesClient;
 
+/**
+ * Variables set by `initialize()`
+ */
+var gStore, unsubscribe, isHighlighted;
+
 function initialize () {
   return Task.spawn(function*() {
     let root = document.querySelector("#app");
-    let store = Store();
+    gStore = Store();
     let app = createElement(App, { front: gFront, heapWorker: gHeapAnalysesClient });
-    let provider = createElement(Provider, { store }, app);
+    let provider = createElement(Provider, { store: gStore }, app);
     render(provider, root);
+    unsubscribe = gStore.subscribe(onStateChange);
   });
 }
 
 function destroy () {
-  return Task.spawn(function*(){});
+  return Task.spawn(function*(){
+    unsubscribe();
+  });
+}
+
+/**
+ * Fired on any state change, currently only handles toggling
+ * the highlighting of the tool when recording allocations.
+ */
+function onStateChange () {
+  let isRecording = gStore.getState().allocations.recording;
+
+  if (isRecording !== isHighlighted) {
+    isRecording ?
+      gToolbox.highlightTool("memory") :
+      gToolbox.unhighlightTool("memory");
+    isHighlighted = isRecording;
+  }
 }
-- 
2.3.1

