From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 27 Oct 2015 08:11:41 -0700
Subject: Bug 1217239 - part 2 - fix bad rebase. r=jlong

diff --git a/devtools/client/memory/components/tree-item.js b/devtools/client/memory/components/tree-item.js
index 990442c..4d92b99 100644
--- a/devtools/client/memory/components/tree-item.js
+++ b/devtools/client/memory/components/tree-item.js
@@ -17,13 +17,13 @@ const TreeItem = module.exports = createClass({
 
     return dom.div({ className: "heap-tree-item" },
       dom.span({ className: "heap-tree-item-bytes" }, item.bytes),
       dom.span({ className: "heap-tree-item-count" }, item.count),
       dom.span({ className: "heap-tree-item-total-bytes" }, item.totalBytes),
       dom.span({ className: "heap-tree-item-total-count" }, item.totalCount),
       dom.span({ className: "heap-tree-item-name", style: { marginLeft: depth * INDENT }},
         arrow,
-        this.toLabel(item.name)
+        item.name
       )
     );
   }
 });
-- 
2.3.1

