From: Jordan Santell <jsantell@mozilla.com>
Date: Thu, 29 Oct 2015 15:47:26 -0700
Subject: Bug 1218673 - Clean up toolbar stlying and be more responsive in the memory tool. r=fitzgen

diff --git a/devtools/client/memory/components/toolbar.js b/devtools/client/memory/components/toolbar.js
index edfcc4d..da537dd 100644
--- a/devtools/client/memory/components/toolbar.js
+++ b/devtools/client/memory/components/toolbar.js
@@ -35,40 +35,42 @@ const Toolbar = module.exports = createClass({
     return (
       dom.div({ className: "devtools-toolbar" },
         dom.button({
           className: `take-snapshot devtools-button`,
           onClick: onTakeSnapshotClick,
           title: L10N.getStr("take-snapshot")
         }),
 
-        dom.label({},
-          L10N.getStr("toolbar.breakdownBy"),
-          dom.select({
-            id: "select-breakdown",
-            className: `select-breakdown`,
-            onChange: e => onBreakdownChange(e.target.value),
-          }, ...breakdowns.map(({ name, displayName }) => dom.option({ key: name, value: name }, displayName)))
-        ),
+        dom.div({ className: "toolbar-group" },
+          dom.label({ className: "breakdown-by" },
+            L10N.getStr("toolbar.breakdownBy"),
+            dom.select({
+              id: "select-breakdown",
+              className: `select-breakdown`,
+              onChange: e => onBreakdownChange(e.target.value),
+            }, ...breakdowns.map(({ name, displayName }) => dom.option({ key: name, value: name }, displayName)))
+          ),
 
-        dom.label({},
-          dom.input({
-            id: "invert-tree-checkbox",
-            type: "checkbox",
-            checked: inverted,
-            onChange: onToggleInverted,
-          }),
-          L10N.getStr("checkbox.invertTree")
-        ),
+          dom.label({},
+            dom.input({
+              id: "invert-tree-checkbox",
+              type: "checkbox",
+              checked: inverted,
+              onChange: onToggleInverted,
+            }),
+            L10N.getStr("checkbox.invertTree")
+          ),
 
-        dom.label({},
-          dom.input({
-            type: "checkbox",
-            checked: allocations.recording,
-            disabled: allocations.togglingInProgress,
-            onChange: onToggleRecordAllocationStacks,
-          }),
-          L10N.getStr("checkbox.recordAllocationStacks")
+          dom.label({},
+            dom.input({
+              type: "checkbox",
+              checked: allocations.recording,
+              disabled: allocations.togglingInProgress,
+              onChange: onToggleRecordAllocationStacks,
+            }),
+            L10N.getStr("checkbox.recordAllocationStacks")
+          )
         )
       )
     );
   }
 });
diff --git a/devtools/client/themes/memory.css b/devtools/client/themes/memory.css
index bc51712..6da3aa4 100644
--- a/devtools/client/themes/memory.css
+++ b/devtools/client/themes/memory.css
@@ -77,16 +77,32 @@ html, .theme-body, #app, #memory-tool, #memory-tool-container {
  */
 #memory-tool .devtools-toolbar {
   display: flex;
   flex-direction: row;
   align-items: center;
   height: var(--toolbar-height);
 }
 
+.devtools-toolbar .toolbar-group {
+  position: absolute;
+  left: 186px;
+  top: 0px;
+  white-space: nowrap;
+}
+
+.toolbar-group > label {
+  margin-right: 5px;
+}
+
+.toolbar-group .breakdown-by span {
+  margin-right: 5px;
+}
+
+
 /**
  * TODO bug 1213100
  * Once we figure out how to store invertable buttons (pseudo element like in this case?)
  * we should add a .invertable class to handle this generally, rather than the definitions
  * in toolbars.inc.css.
  *
  * @see bug 1173397 for another inverted related bug
  */
-- 
2.3.1

