From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 27 Oct 2015 14:49:47 -0700
Subject: Bug 1218675 - Style focused rows in the heap tree correctly. r=fitzgen

diff --git a/devtools/client/memory/components/tree-item.js b/devtools/client/memory/components/tree-item.js
index ac506a8..01f51f0 100644
--- a/devtools/client/memory/components/tree-item.js
+++ b/devtools/client/memory/components/tree-item.js
@@ -14,17 +14,17 @@ const MAX_SOURCE_LENGTH = 200;
  * (▶). When its node has no children, it is hidden.
  */
 const TreeItem = module.exports = createClass({
   displayName: "tree-item",
 
   render() {
     let { item, depth, arrow, focused } = this.props;
 
-    return dom.div({ className: "heap-tree-item" },
+    return dom.div({ className: `heap-tree-item ${focused ? "focused" :""}` },
       dom.span({ className: "heap-tree-item-bytes" }, item.bytes),
       dom.span({ className: "heap-tree-item-count" }, item.count),
       dom.span({ className: "heap-tree-item-total-bytes" }, item.totalBytes),
       dom.span({ className: "heap-tree-item-total-count" }, item.totalCount),
       dom.span({ className: "heap-tree-item-name", style: { marginLeft: depth * INDENT }},
         arrow,
         this.toLabel(item.name)
       )
diff --git a/devtools/client/themes/memory.css b/devtools/client/themes/memory.css
index 31c8810..76703e5 100644
--- a/devtools/client/themes/memory.css
+++ b/devtools/client/themes/memory.css
@@ -115,16 +115,17 @@ html, .theme-body, #app, #memory-tool, #memory-tool-container {
 }
 
 .list > li {
   height: 40px;
   color: var(--theme-body-color);
   border-bottom: 1px solid rgba(128,128,128,0.15);
   padding: 8px;
   cursor: pointer;
+  color: var(--theme-selection-color);
 }
 
 .list > li.selected {
   background-color: var(--theme-selection-background);
   color: var(--theme-selection-color);
 }
 
 .snapshot-list-item span {
@@ -201,18 +202,19 @@ html, .theme-body, #app, #memory-tool, #memory-tool-container {
 .tree-node:nth-child(2n) {
   background-color: var(--row-alt-background-color);
 }
 
 .tree-node:hover {
   background-color: var(--row-hover-background-color);
 }
 
-.tree-node:focus {
+.tree-node:focus, .heap-tree-item.focused {
   background-color: var(--theme-selection-background);
+  color: var(--theme-selection-color);
 }
 
 .header {
   background-color: var(--theme-tab-toolbar-background);
 }
 
 .header span {
   white-space: nowrap;
-- 
2.3.1

