From: Jordan Santell <jsantell@mozilla.com>
Date: Mon, 8 Feb 2016 14:21:42 -0800
Subject: Bug 1244225 - Do not access allocation recording settings when the memory component isn't attached. r=fitzgen

diff --git a/devtools/server/performance/recorder.js b/devtools/server/performance/recorder.js
index 197aade..3c09476 100644
--- a/devtools/server/performance/recorder.js
+++ b/devtools/server/performance/recorder.js
@@ -462,17 +462,23 @@ const PerformanceRecorder = exports.PerformanceRecorder = Class({
   /**
    * Returns the configurations set on underlying components, used in tests.
    * Returns an object with `probability`, `maxLogLength` for allocations, and
    * `features`, `threadFilters`, `entries` and `interval` for profiler.
    *
    * @return {object}
    */
   getConfiguration: function () {
-    return extend({}, this._memory.getAllocationsSettings(), this._profiler.getStartOptions());
+    let allocationSettings = Object.create(null);
+
+    if (this._memory.getState() === "attached") {
+      allocationSettings = this._memory.getAllocationsSettings();
+    }
+
+    return extend({}, allocationSettings, this._profiler.getStartOptions());
   },
 
   toString: () => "[object PerformanceRecorder]"
 });
 
 /**
  * Creates an object of configurations based off of preferences for a PerformanceRecording.
  */
-- 
2.3.1

