From: Jordan Santell <jsantell@mozilla.com>
Date: Tue, 8 Mar 2016 21:58:47 -0800
Subject: Bug 1251033 - Part 3 - Fix unmerged chunk from eb9b84702b2c. a=philor

diff --git a/devtools/client/shared/components/test/mochitest/head.js b/devtools/client/shared/components/test/mochitest/head.js
index 2f960be..f58bce3 100644
--- a/devtools/client/shared/components/test/mochitest/head.js
+++ b/devtools/client/shared/components/test/mochitest/head.js
@@ -134,26 +134,36 @@ var TEST_TREE = {
   expanded: new Set(),
 };
 
 /**
  * Frame
  */
 function checkFrameString (component, file, line, column) {
   let el = component.getDOMNode();
-  is(el.querySelector(".frame-link-filename").textContent, file);
-  is(+el.querySelector(".frame-link-line").textContent, +line);
+  let $ = selector => el.querySelector(selector);
+
+  let $line = $(".frame-link-line");
+  let $column = $(".frame-link-column");
+
+  is($(".frame-link-filename").textContent, file);
+
+  is(el.getAttribute("data-line"), line ? `${line}` : null, "Expected `data-line` found");
+  is(el.getAttribute("data-column"), column ? `${column}` : null, "Expected `data-column` found");
+
+  if (line != null) {
+    is(+$line.textContent, +line);
+  } else {
+    ok(!$line, "Should not have an element for `line`");
+  }
+
   if (column != null) {
-    is(+el.querySelector(".frame-link-column").textContent, +column);
-    is(el.querySelectorAll(".frame-link-colon").length, 2);
+    is(+$column.textContent, +column);
   } else {
-    is(el.querySelector(".frame-link-column"), null,
-      "Should not render column when none specified");
-    is(el.querySelectorAll(".frame-link-colon").length, 1,
-      "Should only render one colon when no column specified");
+    ok(!$column, "Should not have an element for `column`");
   }
 }
 
 function checkFrameTooltips (component, mainTooltip, linkTooltip) {
   let el = component.getDOMNode();
   is(el.getAttribute("title"), mainTooltip);
   is(el.querySelector("a.frame-link-filename").getAttribute("title"), linkTooltip);
 }
-- 
2.3.1

