From: Jordan Santell <jsantell@mozilla.com>
Date: Wed, 23 Mar 2016 14:47:00 -0700
Subject: Bug 1259241 - Update demangling library to 1.0.1 to fix a memory leak. r=fitzgen

diff --git a/devtools/client/shared/demangle.js b/devtools/client/shared/demangle.js
index 1255035..e6792cc 100644
--- a/devtools/client/shared/demangle.js
+++ b/devtools/client/shared/demangle.js
@@ -43,20 +43,22 @@ var ra=[Sc,qc];return{_malloc:vc,_i64Subtract:Cc,_free:wc,_i64Add:Gc,_memmove:Hc
   var m = Module();
   var status = m._malloc(4);
   var buf = m._malloc(2048);
 
   return function(func) {
     if (func.length >= 2048) return null;
     m.writeStringToMemory(func.substr(1), buf);
     var ret = m['___cxa_demangle'](buf, 0, 0, status);
+    var result = null;
     if (m.HEAP32[status >> 2] === 0 && ret) {
-      return m.Pointer_stringify(ret);
+      result = m.Pointer_stringify(ret);
+      m._free(ret);
     }
-    return null;
+    return result;
   };
 })();
 
 // The emscripten compiler exports the Module object; we just want
 // the demangle function
 if (typeof module === "object" && typeof module.exports === "object") {
   module.exports = demangle;
 }
-- 
2.3.1

