From: Jordan Santell <jsantell@gmail.com>
Date: Tue, 17 Feb 2015 18:47:24 -0800
Subject: Bug 980786 - Shader Editor should be disabled when debugging
 chrome. r=vp

diff --git a/browser/devtools/main.js b/browser/devtools/main.js
index 3c9aa84..e65bded 100644
--- a/browser/devtools/main.js
+++ b/browser/devtools/main.js
@@ -214,17 +214,17 @@ Tools.shaderEditor = {
   icon: "chrome://browser/skin/devtools/tool-styleeditor.svg",
   invertIconForLightTheme: true,
   url: "chrome://browser/content/devtools/shadereditor.xul",
   label: l10n("ToolboxShaderEditor.label", shaderEditorStrings),
   panelLabel: l10n("ToolboxShaderEditor.panelLabel", shaderEditorStrings),
   tooltip: l10n("ToolboxShaderEditor.tooltip", shaderEditorStrings),
 
   isTargetSupported: function(target) {
-    return target.hasActor("webgl");
+    return target.hasActor("webgl") && !target.chrome;
   },
 
   build: function(iframeWindow, toolbox) {
     return new ShaderEditorPanel(iframeWindow, toolbox);
   }
 };
 
 Tools.canvasDebugger = {
-- 
2.2.1

