# HG changeset patch
# Parent df2e53ae873088b2b80c975792883eab59d080bb

diff -r df2e53ae8730 testing/config/mozharness/linux_config.py
--- a/testing/config/mozharness/linux_config.py	Wed Aug 27 11:20:23 2014 -0500
+++ b/testing/config/mozharness/linux_config.py	Wed Aug 27 11:36:59 2014 -0500
@@ -12,7 +12,7 @@
         "--extra-profile-file=tests/bin/plugins", "--symbols-path=%(symbols_path)s",
         "--certificate-path=tests/certs", "--autorun", "--close-when-done",
         "--console-level=INFO", "--setpref=webgl.force-enabled=true",
-        "--quiet", "--log-raw=%(raw_log_file)s",
+        "--quiet", "--log-raw=%(raw_log_file)s", "--e10",
         "--use-test-media-devices"
     ],
     "webapprt_options": [
diff -r df2e53ae8730 testing/config/mozharness/mac_config.py
--- a/testing/config/mozharness/mac_config.py	Wed Aug 27 11:20:23 2014 -0500
+++ b/testing/config/mozharness/mac_config.py	Wed Aug 27 11:36:59 2014 -0500
@@ -11,7 +11,7 @@
         "--appname=%(binary_path)s", "--utility-path=tests/bin",
         "--extra-profile-file=tests/bin/plugins", "--symbols-path=%(symbols_path)s",
         "--certificate-path=tests/certs", "--autorun", "--close-when-done",
-        "--console-level=INFO",
+        "--console-level=INFO", "--e10s",
         "--quiet", "--log-raw=%(raw_log_file)s"
     ],
     "webapprt_options": [
diff -r df2e53ae8730 testing/config/mozharness/windows_config.py
--- a/testing/config/mozharness/windows_config.py	Wed Aug 27 11:20:23 2014 -0500
+++ b/testing/config/mozharness/windows_config.py	Wed Aug 27 11:36:59 2014 -0500
@@ -11,7 +11,7 @@
         "--appname=%(binary_path)s", "--utility-path=tests/bin",
         "--extra-profile-file=tests/bin/plugins", "--symbols-path=%(symbols_path)s",
         "--certificate-path=tests/certs", "--autorun", "--close-when-done",
-        "--console-level=INFO",
+        "--console-level=INFO", "--e10s",
         "--quiet", "--log-raw=%(raw_log_file)s"
     ],
     "webapprt_options": [
