From b31156853c3a4b5990351da9eb1cd8e975ee429d Mon Sep 17 00:00:00 2001
From: Jordan Santell <jsantell@mozilla.com>
Date: Fri, 15 May 2015 17:01:24 -0700
Subject: [PATCH] skia test

---
 browser/devtools/performance/test/head.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/browser/devtools/performance/test/head.js b/browser/devtools/performance/test/head.js
index abced57..5e3c637 100644
--- a/browser/devtools/performance/test/head.js
+++ b/browser/devtools/performance/test/head.js
@@ -36,16 +36,18 @@ const IDLE_PREF = "devtools.performance.ui.show-idle-blocks";
 const INVERT_PREF = "devtools.performance.ui.invert-call-tree";
 const INVERT_FLAME_PREF = "devtools.performance.ui.invert-flame-graph";
 const FLATTEN_PREF = "devtools.performance.ui.flatten-tree-recursion";
 const JIT_PREF = "devtools.performance.ui.show-jit-optimizations";
 
 // All tests are asynchronous.
 waitForExplicitFinish();
 
+Services.prefs.setBoolPref("gfx.canvas.azure.accelerated", true);
+
 gDevTools.testing = true;
 
 let DEFAULT_PREFS = [
   "devtools.debugger.log",
   "devtools.performance.ui.invert-call-tree",
   "devtools.performance.ui.flatten-tree-recursion",
   "devtools.performance.ui.show-platform-data",
   "devtools.performance.ui.show-idle-blocks",
-- 
2.2.1

